OpenRA.Mods.Common/Traits/Crates/Crate.cs

```diff
@@ -125,8 +125,39 @@ public void Tick(Actor self)
 		public IEnumerable<Pair<CPos, SubCell>> OccupiedCells() { return new[] { Pair.New(Location, SubCell.FullCell) }; }
 
 		public WPos CenterPosition { get; private set; }
-		public void SetPosition(Actor self, WPos pos) { SetPosition(self, self.World.Map.CellContaining(pos)); }
-		public void SetVisualPosition(Actor self, WPos pos) { SetPosition(self, self.World.Map.CellContaining(pos)); }
+
+		// Sets the location (Location) and visual position (CenterPosition)
+		public void SetPosition(Actor self, WPos pos)
+		{
+			var cell = self.World.Map.CellContaining(pos);
+			SetLocation(self, cell);
+			SetVisualPosition(self, self.World.Map.CenterOfCell(cell) + new WVec(0, 0, pos.Z));
+		}
+
+		// Sets the location (Location) and visual position (CenterPosition)
+		public void SetPosition(Actor self, CPos cell, SubCell subCell = SubCell.Any)
+		{
+			SetLocation(self, cell, subCell);
+			SetVisualPosition(self, self.World.Map.CenterOfCell(cell));
+		}
+
+		// Sets only the visual position (CenterPosition)
+		public void SetVisualPosition(Actor self, WPos pos)
+		{
+			CenterPosition = pos;
+			if (self.IsInWorld)
+			{
+				self.World.ScreenMap.Update(self);
+				self.World.ActorMap.UpdatePosition(self, this);
+			}
+		}
+
+		private void SetLocation(Actor self, CPos cell, SubCell subCell = SubCell.Any)
```

Comment:
I didn't forget, I just thought clarification wasn't necessary with this function. I'll add the comments tho.